apiserver/client/status.go

```diff
@@ -385,6 +406,67 @@ func fetchMachines(st Backend, machineIds set.Strings) (map[string][]*state.Mach
 	return v, nil
 }
 
+// fetchNetworkInterfaces returns maps from machine id to ip.addresses, machine
+// if to a map of interface names from space names, and machine id to
+// linklayerdevices.
+//
+// All are required to determine a machine's network interfaces configuration,
+// so we want all or none.
+func fetchNetworkInterfaces(st Backend) (map[string][]*state.Address, map[string]map[string]set.Strings, map[string][]*state.LinkLayerDevice, error) {
+	ipAddresses := make(map[string][]*state.Address)
+	spaces := make(map[string]map[string]set.Strings)
+	ipAddrs, err := st.AllIPAddresses()
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	for _, ipAddr := range ipAddrs {
+		if !ipAddr.LoopbackConfigMethod() {
+			machineID := ipAddr.MachineID()
+			ipAddresses[machineID] = append(ipAddresses[machineID], ipAddr)
+			subnet, err := st.Subnet(ipAddr.SubnetCIDR())
+			spaceName := ""
+			if errors.IsNotFound(err) {
+				// No worries; no subnets means no spaces.
+			} else if err != nil {
+				return nil, nil, nil, err
+			} else {
+				spaceName = subnet.SpaceName()
+			}
+			if spaceName != "" {
+				deviceName := ipAddr.DeviceName()
+				spacesSet, ok := spaces[machineID][deviceName]
+				if !ok {
+					spaces[machineID] = make(map[string]set.Strings)
```

Comment:
This is going to wipe out the spaces for that machine ID whenever a new device name is seen. I think you want:

```go
devices, ok := spaces[machineID]
if !ok {
    devices = make(map[string]set.Strings)
    spaces[machineID] = devices
}
spacesSet, ok := devices[deviceName]
if !ok {
    spacesSet = make(set.Strings)
    devices[deviceName] = spacesSet
}
```